File: bootstrap/src/mapping.rs
use serde::Serialize;
use std::fmt::{Debug, Formatter};

// type mapping

#[derive(Clone, Serialize)]
pub struct Grammar {
    pub insert: String,
    pub rules: Vec<Rule>,
}

#[derive(Clone, Serialize)]
pub struct Alter {
    pub nameds: Vec<Named>,
    pub inline: String,
}

#[derive(Clone, Serialize)]
pub struct Rule {
    pub name: String,
    pub rstype: String,
    pub alters: Vec<Alter>,
}

#[derive(Clone, Serialize)]
pub enum Named {
    Identifier(String, Atom),
    Anonymous(Atom),
    Cut,
}

#[derive(Clone, Serialize)]
pub enum Atom {
    String(String),
    Name(String),
}

// debug mapping

impl Debug for Grammar {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "\n\"\"\"{}\"\"\"\n{:#?}", self.insert, self.rules)
    }
}

impl Debug for Rule {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}[{}]: {:#?}", self.name, self.rstype, self.alters)
    }
}

impl Debug for Alter {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?} {{ {} }}", self.nameds, self.inline)
    }
}

impl Debug for Named {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Named::Identifier(n, i) => write!(f, "{}={:?}", n, i),
            Named::Anonymous(i) => write!(f, "{:?}", i),
            Named::Cut => write!(f, "~"),
        }
    }
}

impl Debug for Atom {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Atom::String(s) => write!(f, "\"{}\"", s),
            Atom::Name(s) => write!(f, "{}", s),
        }
    }
}

// unwrap cache result mapping

impl From<CacheResult> for Option<Grammar> {
    fn from(value: CacheResult) -> Self {
        match value {
            CacheResult::Grammar(inner) => inner,
            _ => panic!("cache not matched")
        }
    }
}

impl From<CacheResult> for Option<Rule> {
    fn from(value: CacheResult) -> Self {
        match value {
            CacheResult::Rule(inner) => inner,
            _ => panic!("cache not matched")
        }
    }
}

impl From<CacheResult> for Option<Alter> {
    fn from(value: CacheResult) -> Self {
        match value {
            CacheResult::Alter(inner) => inner,
            _ => panic!("cache not matched")
        }
    }
}

impl From<CacheResult> for Option<Named> {
    fn from(value: CacheResult) -> Self {
        match value {
            CacheResult::Named(inner) => inner,
            _ => panic!("cache not matched")
        }
    }
}

impl From<CacheResult> for Option<Atom> {
    fn from(value: CacheResult) -> Self {
        match value {
            CacheResult::Atom(inner) => inner,
            _ => panic!("cache not matched")
        }
    }
}

// cache type mapping

#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug)]
pub enum CacheType {
    Expect(&'static str),
    Grammar,
    String,
    Inline,
    Named,
    Alter,
    Name,
    Rule,
    Atom,
}

// cache result mapping

#[derive(Clone)]
pub enum CacheResult {
    Expect(Option<()>),
    Grammar(Option<Grammar>),
    String(Option<String>),
    Inline(Option<String>),
    Named(Option<Named>),
    Alter(Option<Alter>),
    Name(Option<String>),
    Rule(Option<Rule>),
    Atom(Option<Atom>),
}

// cache result debug mapping

impl Debug for CacheResult {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            CacheResult::Expect(r) => write!(f, "{:?}", r),
            CacheResult::Grammar(r) => write!(f, "{:?}", r),
            CacheResult::String(r) => write!(f, "{:?}", r),
            CacheResult::Inline(r) => write!(f, "{:?}", r),
            CacheResult::Named(r) => write!(f, "{:?}", r),
            CacheResult::Alter(r) => write!(f, "{:?}", r),
            CacheResult::Name(r) => write!(f, "{:?}", r),
            CacheResult::Rule(r) => write!(f, "{:?}", r),
            CacheResult::Atom(r) => write!(f, "{:?}", r),
        }
    }
}

File: templates/mini/mapping.py
from typing import IO
from templates.shared import CLAIM, Generator


class Mapping(Generator):
    __prelude = '''
use serde::Serialize;
use std::fmt::{Debug, Formatter};
'''

    def __init__(self, peg, file: IO[str] | None = None) -> None:
        super().__init__(peg, file)

    def generate(self) -> None:
        self.print(CLAIM)
        self.print(self.__prelude)
        self.print(self.json['insert'])
        for each in self.node:
            self.print(f'impl From<CacheResult> for Option<{each}> {{')
            with self.indent():
                self.print('fn from(value: CacheResult) -> Self {')
                with self.indent():
                    self.print('match value {')
                    with self.indent():
                        self.print(f'CacheResult::{each}(inner) => inner,')
                        self.print('_ => panic!("cache not matched")')
                    self.print('}')
                self.print('}')
            self.print('}')
            self.print()

        self.print('#[derive(Eq, PartialEq, Hash, Clone, Copy, Debug)]')
        self.print('pub enum CacheType {')
        with self.indent():
            for each in ['Expect(&\'static str)', 'String', 'Inline', 'Name'] + self.node:
                self.print(f'{each},')
        self.print('}')
        self.print()
        self.print('#[derive(Clone)]')
        self.print('pub enum CacheResult {')
        with self.indent():
            self.print('Expect(Option<()>),')
            self.print('String(Option<String>),')
            self.print('Inline(Option<String>),')
            self.print('Name(Option<String>),')
            for each in self.node:
                self.print(f'{each}(Option<{each}>),')
        self.print('}')
        self.print()
        self.print('impl Debug for CacheResult {')
        with self.indent():
            self.print('fn fmt(&self, f: &mut Formatter<\'_>) -> std::fmt::Result {')
            with self.indent():
                self.print('match self {')
                with self.indent():
                    for each in ['Expect', 'String', 'Inline', 'Name'] + self.node:
                        self.print(f'CacheResult::{each}(r) => write!(f, "{{:?}}", r),')
                self.print('}')
            self.print('}')
        self.print('}')
